core/opendaq/modulemanager/src/module_manager_impl.cpp

```diff
@@ -94,6 +96,250 @@ ErrCode ModuleManagerImpl::loadModules(IContext* context)
+            availableDevicesPtr.pushBack(deviceInfo);
+    }
+
+    *availableDevices = availableDevicesPtr.detach();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode ModuleManagerImpl::getAvailableDeviceTypes(IDict** deviceTypes)
+{
+    OPENDAQ_PARAM_NOT_NULL(deviceTypes);
+    
+    auto availableTypes = Dict<IString, IDeviceType>();
+
+    for (const auto& library : libraries)
+    {
+        const auto module = library.module;
+
+        DictPtr<IString, IDeviceType> moduleDeviceTypes;
+
+        try
+        {
+            moduleDeviceTypes = module.getAvailableDeviceTypes();
+        }
+        catch (NotImplementedException&)
+        {
+            LOG_I("{}: GetAvailableDeviceTypes not implemented", module.getName())
+        }
+        catch ([[maybe_unused]] const std::exception& e)
+        {
+            LOG_W("{}: GetAvailableDeviceTypes failed: {}", module.getName(), e.what())
+        }
+
+        if (!moduleDeviceTypes.assigned())
+            continue;
+
+        for (const auto& [id, type] : moduleDeviceTypes)
+            availableTypes.set(id, type);
+    }
+
+    *deviceTypes = availableTypes.detach();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode ModuleManagerImpl::createDevice(IDevice** device, IString* connectionString, IComponent* parent, IPropertyObject* config)
+{
+    OPENDAQ_PARAM_NOT_NULL(device);
+
+    for (const auto& library : libraries)
+    {
+        const auto module = library.module;
+
+        bool accepted{};
+        try
+        {
+            accepted = module.acceptsConnectionParameters(connectionString, config);
+        }
+        catch (NotImplementedException&)
+        {
+            LOG_I("{}: AcceptsConnectionString not implemented", module.getName())
+            accepted = false;
+        }
+        catch ([[maybe_unused]] const std::exception& e)
+        {
+            LOG_W("{}: AcceptsConnectionString failed: {}", module.getName(), e.what())
+            accepted = false;
+        }
+
+        if (accepted)
+        {
+            return module->createDevice(device, connectionString, parent, config);
+        }
+    }
+
+    return this->makeErrorInfo(
+        OPENDAQ_ERR_NOTFOUND, "Device with given connection string and config is not available [{}]", connectionString);
+}
+
+ErrCode ModuleManagerImpl::getAvailableFunctionBlockTypes(IDict** functionBlockTypes)
+{
+    OPENDAQ_PARAM_NOT_NULL(functionBlockTypes);
+
+    auto availableTypes = Dict<IString, IFunctionBlockType>();
+
+    for (const auto& library : libraries)
+    {
+        const auto module = library.module;
+        
+        DictPtr<IString, IFunctionBlockType> types;
+        try
+        {
+            types = module.getAvailableFunctionBlockTypes();
+        }
+        catch (NotImplementedException&)
+        {
+            LOG_I("{}: GetAvailableFunctionBlockTypes not implemented", module.getName())
+        }
+        catch ([[maybe_unused]] const std::exception& e)
```

Comment:
You're actually not using `e` so comment out the name `/*e*/` and you can remove `[[maybe_unused]]`